Source/Common/Profiler.h

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: MIT
+/*
+$info$
+tags: Common|Profiler
+desc: Frontend profiler common code
+$end_info$
+*/
+#pragma once
+#include <FEXCore/Utils/Profiler.h>
+
+namespace FEXCore::Core {
+struct InternalThreadState;
+}
+
+#ifdef _M_ARM_64
+static inline void memory_barrier() {
+  asm volatile("dmb ishst;" ::: "memory");
+}
+
+#else
+static inline void memory_barrier() {
+  // Intentionally empty.
+}
+#endif
+
+namespace FEX::Profiler {
+class StatAllocBase {
+public:
+  virtual ~StatAllocBase() = default;
+
+protected:
+  FEXCore::Profiler::ThreadStats* AllocateBaseSlot(uint32_t TID);
+  void DeallocateBaseSlot(FEXCore::Profiler::ThreadStats* AllocatedSlot);
+
+  uint32_t OffsetFromStat(FEXCore::Profiler::ThreadStats* Stat) const {
+    return reinterpret_cast<uint64_t>(Stat) - reinterpret_cast<uint64_t>(Base);
+  }
+  size_t TotalSlotsFromSize() const {
+    return (CurrentSize - sizeof(FEXCore::Profiler::ThreadStatsHeader)) / sizeof(FEXCore::Profiler::ThreadStats) - 1;
+  }
+  size_t SlotIndexFromOffset(uint32_t Offset) {
+    return (Offset - sizeof(FEXCore::Profiler::ThreadStatsHeader)) / sizeof(FEXCore::Profiler::ThreadStats);
+  }
+
+  void SaveHeader(FEXCore::Profiler::AppType AppType);
+
+  void* Base;
+  size_t CurrentSize {};
+  FEXCore::Profiler::ThreadStatsHeader* Head {};
+  FEXCore::Profiler::ThreadStats* Stats;
+  FEXCore::Profiler::ThreadStats* StatTail {};
```

Comment:
Why are you preferring a custom-implemented linked list over flat memory? This seems neither more convenient for the implementation nor like it would improve performance.

Reply:
Sparsity in the allocation buffer is a concern, so this is how I'm keeping the implementation.

Reply:
Could you elaborate what exactly the problem is?

Reply:
What scenario exactly would trigger this? I don't know what your reader side looks like, but it sounds like a bitmask and/or active slot count would easily resolve this.

Reply:
I'm just going to leave it as a linked list. We can rev the version number and change it to a bitmask in the future if someone is feeling up to writing that code in the future.

Reply:
If the choice between linked lists and flat memory doesn't seem to be so critical, why is the focus on weakly-ordered atomics considered crucial?

Reply:
For the atomics I don't want the writer side to be affected by cacheline sharing problems at all. Semantically these are two independent things. The containing data which is `FEXCore::Profiler::ThreadStats` and the way to walk the list of data.

Reply:
I understand they are different things, but it seems arbitrary to want this (with no explicit motivation) while using a data structure with poor performance characteristics in the same module.

Reply:
Indeed, which is why we leave some of the "poor performance" problems on the reader side which has something like 500ms to sample the data, even though this is going to be hundreds of nanoseconds regardless.